src/Cardano/Prelude/GHC/Heap/Tree.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{-|
+Module      :  Cardano.Prelude.GHC.Heap.Tree
+
+This module exports functions and data types for building and printing GHC
+heap object trees. Very useful for debugging issues involving the heap
+representation of Haskell expressions.
+
+-}
+
+module Cardano.Prelude.GHC.Heap.Tree
+  ( ClosureTreeOptions(..)
+  , RecurseCyclicClosures(..)
+  , TreeDepth(..)
+  , buildClosureTree
+  , buildAndRenderClosureTree
+  , depth
+  , renderClosure
+  , renderTree
+  )
+where
+
+import Cardano.Prelude.Base
+
+import Data.Text (pack, unpack)
+import Data.Tree (Tree(..), drawTree, levels)
+import GHC.Exts.Heap
+  (Box(..), Closure, allClosures, areBoxesEqual, asBox, getClosureData)
+import System.Mem (performGC)
+
+-- | The depth of a 'Tree'.
+data TreeDepth
+  = TreeDepth {-# UNPACK #-} !Int
+  -- ^ A specific tree depth bound.
+  | AnyDepth
+  -- ^ No tree depth bound.
+  deriving (Show)
+
+-- | Whether to recurse cyclic closures in a 'Closure' 'Tree'.
+data RecurseCyclicClosures
+  = RecurseCyclicClosures
+  -- ^ Recurse cyclic closures.
+  | NoRecurseCyclicClosures
```

Comment:
Minor wording quibble, but I might say `TraverseCyclicClosures`.

Reply:
Also, I'm curious - in the case of `RecurseCyclicClosures`, how do we avoid looping forever?

Reply:
@mhuesch:
> Minor wording quibble, but I might say `TraverseCyclicClosures`.
Cool, I agree with this!

> Also, I'm curious - in the case of RecurseCyclicClosures, how do we avoid looping forever?
You would need to specify a max `TreeDepth` in the `ClosureTreeOptions` to prevent an infinite loop in this case.